### PR TITLE
stdcm: use spacing requirements instead of route occupancies

### LINKED ISSUE #5148: meta: stdcm: use block occupancies instead of route occupancies
Requires https://github.com/osrd-project/osrd/pull/5128 to be merged.

- [x] #5149
- [x] #5150
- [x] #5151

There's several issues to be split between different teams, but it should all be merged into a single branch first. Individual changes would break the app.

### LINKED ISSUE #5149: stdcm: editoast: use spacing requirements instead of route occupancies in the request sent to core

### LINKED ISSUE #5150: stdcm: core: use spacing requirement data to build block occupancies
- [ ] Change the STDCM request format to include spacing requirement data instead of route occupancies
  - [ ] Adapt the openapi file accordingly
- [ ] Update UnavailableSpaceBuilder

### LINKED ISSUE #5151: stdcm: editoast: remove route occupancies from `ResultTrain`

### PR DESCRIPTION
**Pull request by eckter**
- [x] Use spacing requirements
- [x] Update test values
- [x] Update openapi file

The zone ids used for resource usage is now a property of the zone, it's still generated from its detectors.

Fixes #5148.
Fixes #5149.
Fixes #5150.
Fixes #5151.

### CODE REVIEW

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt [hunk 1]
```diff
@@ -109,7 +111,7 @@ fun RoutingInfra.getRouteEntry(route: RouteId): DirDetectorId {
 }
 
 @JvmName("getRouteExit")
```
Comment (Erashin) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt]: `getBlockExit`.

File: core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java [hunk 2]
```diff
@@ -29,120 +31,95 @@ public class UnavailableSpaceBuilder {
...
+                        var secPreBlockLength = blockInfra.getBlockLength(secondPredecessorBlock);
+                        unavailableSpace.put(secondPredecessorBlock, new OccupancySegment(
+                                timeRange.lowerEndpoint(), timeRange.upperEndpoint(),
+                                Math.max(0, secPreBlockLength - SIGHT_DISTANCE), preBlockLength
+                        ));
+                    }
+                }
+
+                // Generate train length occupancy
+                var successorBlocks = getNextBlocks(infra, blockInfra, blockId);
```
Comment (Erashin) [inline core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java:76]: What if the train's length is > to the next block's length ?  Shouldn't the block after that be marked as used, for `length = min(nextNextBlockLength, Distance.fromMeters(rollingStock.getLength()) - nextBlockLength)` ?

**Review by Erashin — CHANGES_REQUESTED** [review]

File: core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java (hunk 2 above)
Comment (eckter) [reply core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java:76]: True, it should be. But I don't know if it's worth the added code complexity, this code is meant to be replaced very soon. It's an uncommon case and doesn't cause significant issues.

**Review by Erashin — APPROVED** [review]
LGTM

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt [hunk 3]
```diff
@@ -108,8 +110,8 @@ fun RoutingInfra.getRouteEntry(route: RouteId): DirDetectorId {
     return getZonePathEntry(getRoutePath(route).first())
 }
 
-@JvmName("getRouteExit")
-fun RoutingInfra.getRouteExit(route: RouteId): DirDetectorId {
+@JvmName("getBlockExit")
+fun RoutingInfra.getBlockExit(route: RouteId): DirDetectorId {
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt]: This rename must be a mistake. getRouteExit gets the exit detector or a route, blocks have nothing to do with it

**Review by multun — CHANGES_REQUESTED** [review]

File: core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java [hunk 4]
```diff
@@ -29,120 +31,95 @@ public class UnavailableSpaceBuilder {
...
+        return new HashSet<>(toIntList(blockInfra.getBlocksEndingAtDetector(entry)));
     }
 
-    /** Returns the routes that follow the given one */
-    private static Set<Integer> getNextRoutes(RawSignalingInfra infra, int route) {
-        var exit = InterlockingInfraKt.getRouteExit(infra, route);
-        return new HashSet<>(toIntList(infra.getRoutesStartingAtDet(exit)));
+    /** Returns the blocks that follow the given one */
+    private static Set<Integer> getNextBlocks(RawSignalingInfra infra, BlockInfra blockInfra, int blockId) {
+        var entry = BlockInfraImplKt.getBlockExit(blockInfra, infra, blockId);
```
Comment (multun) [inline core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java:122]: this is a type confusion bug: getBlockExit is in fact getRouteExit, and was renamed for some reason, so it takes a routeId

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt [hunk 5]
```diff
@@ -108,8 +110,8 @@ fun RoutingInfra.getRouteEntry(route: RouteId): DirDetectorId {
...
     return getZonePathExit(getRoutePath(route).last())
 }
```
Comment (Khoyo) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt]: What?

**Review by Khoyo — CHANGES_REQUESTED** [review]

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt (hunk 3 above)
Comment (eckter) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/InterlockingInfra.kt]: Definitely a mistake, I must have accidentally used a fancy IDE feature.

@Erashin did catch the name clash but I fixed it the wrong way around...

File: core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java (hunk 4 above)
Comment (eckter) [reply core/src/main/java/fr/sncf/osrd/stdcm/preprocessing/implementation/UnavailableSpaceBuilder.java:122]: This is actually the correct call, it calls the `BlockInfra` version which does get the exit detector of a block.

The accidental rename of `getRouteExit` has surprisingly nothing to do with this.

**Review by Khoyo — APPROVED** [review]

**Review by multun — APPROVED** [review]

**Review by flomonster — APPROVED** [review]
Reviewed editoast

**Pull request merged**